metricbeat/module/postgresql/statement/statement.go

```diff
@@ -44,33 +41,25 @@ func init() {
 // mb.BaseMetricSet because it implements all of the required mb.MetricSet
 // interface methods except for Fetch.
 type MetricSet struct {
-	mb.BaseMetricSet
+	*postgresql.MetricSet
 }
 
 // New creates a new instance of the MetricSet. New is responsible for unpacking
 // any MetricSet specific configuration options if there are any.
 func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
-	config := struct{}{}
-	if err := base.Module().UnpackConfig(&config); err != nil {
+	ms, err := postgresql.NewMetricSet(base)
+	if err != nil {
 		return nil, err
 	}
-
-	return &MetricSet{
-		BaseMetricSet: base,
-	}, nil
+	return &MetricSet{MetricSet: ms}, nil
 }
 
 // Fetch methods implements the data gathering and data conversion to the right
 // format. It publishes the event which is then forwarded to the output. In case
 // of an error set the Error field of mb.Event or simply call report.Error().
 func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
-	db, err := sql.Open("postgres", m.HostData().URI)
-	if err != nil {
-		return errors.Wrap(err, "error in Open")
-	}
-	defer db.Close()
-
-	results, err := postgresql.QueryStats(db, "SELECT * FROM pg_stat_statements")
+	ctx := context.Background()
+	results, err := m.QueryStats(ctx, "SELECT * FROM pg_stat_statements")
```

Comment:
I'm guessing metricset `Close` won't cancel these requests, I guess it would be unexpected for them to leak, but it could be worth it to handle ctx cancellation?

Reply:
Right, `Close` doesn't cancel running queries, they will be executed till they finish. I wouldn't expect leaks because of that.
Not sure if it would worth to hande cancelation here, we'd need to keep a thread safe list of cancelation functions available on close. Once we have #11981 or similar we will be able to cancel the parent context when calling the closers.

Reply:
++ I thought about that PR too. That said, you don't really need to keep a list, isn't it? When creating the metricset you can create a Background context, use it in all fetch operations and cancel it on close.

Reply:
Having that that PR is moving (:crossed_fingers:) I'm ok with leaving this for update after it's merged

Reply:
> That said, you don't really need to keep a list, isn't it? When creating the metricset you can create a Background context, use it in all fetch operations and cancel it on close.

Yes, this would be another option, but I think this is what should be done in a general what by #11981. Doing it this way here has some disadvantages, we'd be hiding the context inside the metricset and it would be a different one to the one received by `Fetch` if #11981 is implemented.

> Having that that PR is moving (crossed_fingers) I'm ok with leaving this for update after it's merged

Yeah, let me move this forward :slightly_smiling_face: